### PR TITLE
cql3: functions: validate arguments for 'token()' also

### LINKED ISSUE #10448: The CQL token() function should validate its parameters
As [our documentation of the `token()` function](https://docs.scylladb.com/getting-started/functions/) explains,

> The arguments of the token depend on the type of the partition key columns that are used. 

If the number or type of arguments passed to `token()`'s do not match the requirements of the partition key, an invalid request error should be thrown. Currently it is not - the SELECT seems to silently match nothing but I'm worried that under the hood it might be doing something undefined and worse could happen - as happened in Cassandra which had the same bug (https://issues.apache.org/jira/browse/CASSANDRA-14989) - and had a null pointer exception in this case. The Cassandra bug was fixed in Cassandra 4.0.

This bug is reproduced by two unit tests translated from Cassandra:
* `cassandra_tests/validation/operations/select_test.py`::`testTokenFctRejectsInvalidColumnType`
* `cassandra_tests/validation/operations/select_test.py`::`testTokenFctRejectsInvalidColumnCount`

### PR DESCRIPTION
**Pull request by tchaikov**
since "token()" computes the token for a given partition key, if we pass the key of the wrong type, it should reject.

in this change, we validate the keys before returning the "token()" function.

Fixes #10448
Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
see also #12677

hi @cvybhu hi Jan, just wanted to help out, but not sure if this is the right fix. could you take a look?

**Comment by tchaikov** [comment]
regarding the test failure, shall we be compatible with a higher version of Cassandra, see https://issues.apache.org/jira/browse/CASSANDRA-14989 and https://github.com/apache/cassandra/commit/174cf761f7897443080b8a840b649b7eab17ae25

where `token()` rejects the parameter set if its number is wrong. please note, in the original issue #10594, we were testing "Cassandra 3.11.8 (and Cassandra 2.2.0)". and the above cassandra commit was included in Cassandra 4.0 and up.

**Comment by nyh** [comment]
> regarding the test failure, shall we be compatible with a higher version of Cassandra? see https://issues.apache.org/jira/browse/CASSANDRA-14989 and [apache/cassandra@174cf76](https://github.com/apache/cassandra/commit/174cf761f7897443080b8a840b649b7eab17ae25)
> 
> where `token()` rejects the pack of parameters if its size is wrong. please note, in the original issue #10594, we were testing "Cassandra 3.11.8 (and Cassandra 2.2.0)". and the above cassandra commit was included in Cassandra 4.0 and up. @avelanarius what do you think?

I don't understand the question. The test failure reported in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4602/testReport/junit/(root)/non-boost%20tests/cql_token_from_null_test_dev_1/ is a "test/cql", whose problem in the first place (and why I don't personally use this test framework) is that those tests were never actually tested against Cassandra - not 2, not 3 and not 4. Issue #10594 was about token(v) with the correct number of arguments but it was null. It seems the test shows it still works. The test also tested the wrong number of arguments and expected it to work - this makes no sense to me, and I don't think it ever worked in Cassandra.

I think this is a test bug and should be fixed - in the same patch where you fix the code.

**Comment by tchaikov** [comment]
> > regarding the test failure, shall we be compatible with a higher version of Cassandra? see https://issues.apache.org/jira/browse/CASSANDRA-14989 and [apache/cassandra@174cf76](https://github.com/apache/cassandra/commit/174cf761f7897443080b8a840b649b7eab17ae25)
> > where `token()` rejects the pack of parameters if its size is wrong. please note, in the original issue #10594, we were testing "Cassandra 3.11.8 (and Cassandra 2.2.0)". and the above cassandra commit was included in Cassandra 4.0 and up. @avelanarius what do you think?
> 
> I don't understand the question. 

the question is, "shall we be compatible with a higher version of Cassandra?" i think the behavior exists in an older version of Cassandra. but it was changed since Cassandra 4.0.

> The test failure reported in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4602/testReport/junit/(root)/non-boost%20tests/cql_token_from_null_test_dev_1/ is a "test/cql", whose problem in the first place (and why I don't personally use this test framework) is that those tests were never actually tested against Cassandra - not 2, not 3 and not 4. Issue #10594 was about token(v) with the correct number of arguments but it was null. It seems the test shows it still works. 

yes.

> The test also tested the wrong number of arguments and expected it to work - this makes no sense to me, and I don't think it ever worked in Cassandra.
> 

what's what changed since Cassandra 4.0 since the commit i referenced above.

> I think this is a test bug and should be fixed - in the same patch where you fix the code.

anyhow, i think the new behavior makes more sense so changed accordingly.

**Comment by nyh** [comment]
> the question is, "shall we be compatible with a higher version of Cassandra?" i think the behavior exists in an older version of Cassandra. but it was changed since Cassandra 4.0.

I didn't see any proof that "the behavior" exhibited by that test/cql/test - namely that you can just pass 3 parameters to token() instead of 1 and it will silently ignore the other two - ever existed in Cassandra. You pointed to a **bug** in Cassandra regarding multi-component partition keys, but this is not what this test is testing (unless I misunderstood something...).

Only if you are convinced that both Cassandra 3 and 4 had sensible non-buggy behavior on something, but they were different behaviors - only then we can ask the question of which one to be compatible with. But I don't think this is the case here.

Anyway, this discussion is one of the reason I think it is important to be able to run our tests against Cassandra. test/cql-pytest already can do that from day one,  test/cql could do it in theory but nobody did it (https://github.com/scylladb/scylladb/issues/10098).

> 
> > The test also tested the wrong number of arguments and expected it to work - this makes no sense to me, and I don't think it ever worked in Cassandra.
> 
> that's what changed since Cassandra 4.0 since the commit i referenced above.

This is not how I understood the commit. In any case, instead of trying to understand the commit, we can run Cassandra and check. Did you see that Cassandra 3 silently returned null instead of erroring out (or crashing!) when you try to pass 3 parameters to token() instead of just one? If that's not you saw, then the test isn't compatible with Cassandra 3 and this is all a red herring - and the test is just buggy and we never knew it because we never did https://github.com/scylladb/scylladb/issues/10098.

> > I think this is a test bug and should be fixed - in the same patch where you fix the code.
> anyhow, i think the new behavior makes more sense so changed accordingly.

I agree. But if you believe that the old behavior was compatible with Cassandra 3, it's a theory you should check - if true the commit message and test should be clear about this change - if it's false then we shouldn't suggest this at all, maybe it was just a simple test bug that needs to be fixed.

**Review by nyh — CHANGES_REQUESTED** [review]
I see you un-xfailed testTokenFctRejectsInvalidColumnType, but a second test - testTokenFctRejectsInvalidColumnCount should have passed as well (issue #10448 mentions both these tests). I am guessing it didn't pass because of slightly different error messages between Scylla and Cassandra, so please fix the error message checked (e.g., remove a few words, etc.)  so the test will pass and can be un-xfailed too. Thanks.

**Comment by tchaikov** [comment]
> I see you un-xfailed testTokenFctRejectsInvalidColumnType, but a second test - testTokenFctRejectsInvalidColumnCount should have passed as well (issue #10448 mentions both these tests). 
> I am guessing it didn't pass because of slightly different error messages between Scylla and Cassandra, 

no. that's not related to the "slightly different error messages between Scylla and Cassandra". as the commit message puts, before my change, the parameters passed  "token()" function's were not validated at all. 

if i make the change based my guess, i will surely put it clearly in the commit message.

**Comment by nyh** [comment]
> > I see you un-xfailed testTokenFctRejectsInvalidColumnType, but a second test - testTokenFctRejectsInvalidColumnCount should have passed as well (issue #10448 mentions both these tests).
> > I am guessing it didn't pass because of slightly different error messages between Scylla and Cassandra,
> 
> no. that's not related to the "slightly different error messages between Scylla and Cassandra". as the commit message puts, before my change, the parameters passed "token()" function's were not validated at all.

I don't understand... Why doesn't the test `testTokenFctRejectsInvalidColumnCount` begin to pass after your patch? What error happens? You can run the single test as:

```
test/cql-pytest/run --runxfail cassandra_tests/validation/operations/select_test.py::testTokenFctRejectsInvalidColumnCount & less
```

In any case, you can't have your patch "Fixes 10448" and yet leave one failing test that claims it xfails because of 10448... You should modify either Scylla or the test to fix this second test as well - or, if your patch only solves part of the problem and leaves some sort of new/different/partial issue open, you should open a new issue on the remaining problem and change the test to claim it xfails because of this new issue instead of the already-fixed 10448.

**Comment by tchaikov** [comment]
> > the question is, "shall we be compatible with a higher version of Cassandra?" i think the behavior exists in an older version of Cassandra. but it was changed since Cassandra 4.0.
> 
> I didn't see any proof that "the behavior" exhibited by that test/cql/test - namely that you can just pass 3 parameters to token() instead of 1 and it will silently ignore the other two - ever existed in Cassandra. 

agreed. i think i just blindly trusted the change which introduced this change without running them against Cassandra.

> You pointed to a **bug** in Cassandra regarding multi-component partition keys, but this is not what this test is testing (unless I misunderstood something...).
> 

right. 

1. create a table with a pkey of 2 component
2. select `token(cart_id)`, and Cassandra throws `NullPointerException`

but what we are testing is to feed the function with more number of parameters. these tests are not covered by https://github.com/apache/cassandra/commit/174cf761f7897443080b8a840b649b7eab17ae25 . it does cover the case where a pk with 2 component, but only a single parameter is passed to token though.

yeah, i think the referenced jira ticket is relevant but not exactly the one i was looking for.

> Only if you are convinced that both Cassandra 3 and 4 had sensible non-buggy behavior on something, but they were different behaviors - only then we can ask the question of which one to be compatible with. But I don't think this is the case here.
> 

correct, not yet. i misunderstood the jira ticket.

> Anyway, this discussion is one of the reason I think it is important to be able to run our tests against Cassandra. test/cql-pytest already can do that from day one, test/cql could do it in theory but nobody did it (#10098).
> 
> > > The test also tested the wrong number of arguments and expected it to work - this makes no sense to me, and I don't think it ever worked in Cassandra.
> > 
> > 
> > that's what changed since Cassandra 4.0 since the commit i referenced above.
> 
> This is not how I understood the commit. In any case, instead of trying to understand the commit, we can run Cassandra and check. Did you see that Cassandra 3 silently returned null instead of erroring out (or crashing!) when you try to pass 3 parameters to token() instead of just one? If that's not you saw, then the test isn't compatible with Cassandra 3 and this is all a red herring - and the test is just buggy and we never knew it because we never did #10098.

okay. seems i have to run Cassandra locally. was not really in favor of running a server in Java, but the day has come anyway.

> 
> > > I think this is a test bug and should be fixed - in the same patch where you fix the code.
> > > anyhow, i think the new behavior makes more sense so changed accordingly.
> 
> I agree. But if you believe that the old behavior was compatible with Cassandra 3, it's a theory you should check - if true the commit message and test should be clear about this change - if it's false then we shouldn't suggest this at all, maybe it was just a simple test bug that needs to be fixed.

no, i misinterpret the jira ticket. i don't support this theory anymore.

**Comment by mykaul** [comment]
> okay. seems i have to run Cassandra locally. was not really in favor of running a server in Java, but the day has come anyway.

Try https://www.testcontainers.org/ - there's a Cassandra one there (wish there was Scylla too!)

**Comment by tchaikov** [comment]
> > > I see you un-xfailed testTokenFctRejectsInvalidColumnType, but a second test - testTokenFctRejectsInvalidColumnCount should have passed as well (issue #10448 mentions both these tests).
> > > I am guessing it didn't pass because of slightly different error messages between Scylla and Cassandra,
> > 
> > 
> > no. that's not related to the "slightly different error messages between Scylla and Cassandra". as the commit message puts, before my change, the parameters passed "token()" function's were not validated at all.
> 
> I don't understand... Why doesn't the test `testTokenFctRejectsInvalidColumnCount` begin to pass after your patch? What error happens? You can run the single test as:
> 

ahh! sorry, i missed that one! should have grep'ed the source or looked around.

> ```
> test/cql-pytest/run --runxfail cassandra_tests/validation/operations/select_test.py::testTokenFctRejectsInvalidColumnCount & less
> ```

it passed. lemme drop the xfail decorator.

> 
> In any case, you can't have your patch "Fixes 10448" and yet leave one failing test that claims it xfails because of 10448... You should modify either Scylla or the test to fix this second test as well - or, if your patch only solves part of the problem and leaves some sort of new/different/partial issue open, you should open a new issue on the remaining problem and change the test to claim it xfails because of this new issue instead of the already-fixed 10448.

indeed. will update and repush. thanks!

**Comment by tchaikov** [comment]
> > okay. seems i have to run Cassandra locally. was not really in favor of running a server in Java, but the day has come anyway.
> 
> Try https://www.testcontainers.org/ - there's a Cassandra one there (wish there was Scylla too!)

@mykaul hey Yaniv, thank you! you saved my day. will try it out early tomorrow.

**Comment by nyh** [comment]
> okay. seems i have to run Cassandra locally. was not really in favor of running a server in Java, but the day has come anyway.

‎By the way, this is one of the goals of the cql-pytest framework - to help you run Cassandra locally very easily. Just like test/cql-pytest/run can run cql-pytest tests against the latest-compiled Scylla, test/cql-pytest/run-cassandra helps you run the test against Cassandra. You just need to download any version of Cassandra you want, build it (fairly trivial), and set the CASSANDRA variable to tell test/cql-pytest/run-cassandra to run it for you. You only need to do this once and then use the same Cassandra for a year (or whatever).

**Comment by nyh** [comment]
By the way, @tchaikov take a look at `test/cql-pytest/README.md` for an explanation of this framework's philosophy, including the desire - and ability - to run tests against Cassandra. 
I know that the buggy test you found here wasn't a test/cql-pytest but rather test/cql, which doesn't make it easy to run the test against Cassandra, but that should be fixed (https://github.com/scylladb/scylladb/issues/10098).

**Review by avikivity — APPROVED** [review]

**Comment by tchaikov** [comment]
@nyh hi Nadav, that'd be very helpful! thank you! will read through them tomorrow. i guess i just don't have the courage to face the pile of XML + jar . but with the nice test framework and dock images, i should be able to containerize (pun intended) the fear =)

**Comment by nyh** [comment]
> @nyh hi Nadav, that'd be very helpful! thank you! will read through them tomorrow. i guess i just don't have the courage to face the pile of XML + jar .

It's not that difficult, it's much easier than running Scylla. You can download Cassandra source code and just run the build command specified in the README, and then you can either run it with bin/cassandra or if you write cql-pytest use the test/cql-pytest/run-cassandra tool. You can also download an already-compile Cassandra binary. Or, I guess you can also use containers...

In any case, you don't really need to run Cassandra - I only suggested that you need to run Cassandra because you made some wrong (I think) claims about what Cassandra 3 did. You don't need to run Cassandra to fix this patch, you just need to un-xfail the last test - a test that I already verified that works on Cassandra (in fact, it was a Cassandra unit test), so you don't really need to do it yourself if you don't want to.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
This version looks good. Let's just wait for the CI now before merging. Thanks!

**Pull request closed without merging**
